incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -2244,22 +2244,74 @@ private void onDecodeDisconnect(
             long authorization,
             MqttDisconnectFW disconnect)
         {
-            state = MqttState.closingInitial(state);
-            if (session)
+            byte reasonCode = decodeDisconnectProperties(disconnect.properties());
+
+            if (reasonCode != SUCCESS)
             {
-                if (disconnect.reasonCode() == DISCONNECT_WITH_WILL_MESSAGE)
-                {
-                    sessionStream.doSessionAbort(traceId);
-                }
-                else
+                onDecodeError(traceId, authorization, reasonCode);
+                decoder = decodeIgnoreAll;
+            }
+            else
+            {
+                if (session)
                 {
-                    sessionStream.doSessionAppEnd(traceId, EMPTY_OCTETS);
+                    if (disconnect.reasonCode() == DISCONNECT_WITH_WILL_MESSAGE)
+                    {
+                        sessionStream.doSessionAbort(traceId);
+                    }
+                    else
+                    {
+                        sessionStream.doSessionAppEnd(traceId, EMPTY_OCTETS);
+                    }
                 }
             }
+
+            state = MqttState.closingInitial(state);
             closeStreams(traceId, authorization);
             doNetworkEnd(traceId, authorization);
         }
 
+        private byte decodeDisconnectProperties(
+            MqttPropertiesFW properties)
+        {
+            byte reasonCode = SUCCESS;
+
+            final OctetsFW propertiesValue = properties.value();
+            final DirectBuffer decodeBuffer = propertiesValue.buffer();
+            final int decodeOffset = propertiesValue.offset();
+            final int decodeLimit = propertiesValue.limit();
+
+            decode:
+            for (int decodeProgress = decodeOffset; decodeProgress < decodeLimit; )
+            {
+                final MqttPropertyFW mqttProperty = mqttPropertyRO.wrap(decodeBuffer, decodeProgress, decodeLimit);
+                switch (mqttProperty.kind())
+                {
+                case KIND_SESSION_EXPIRY:
+                    if (isSetSessionExpiryInterval(propertyMask))
+                    {
+                        reasonCode = PROTOCOL_ERROR;
+                        break decode;
+                    }
+                    this.propertyMask |= CONNECT_SESSION_EXPIRY_INTERVAL_MASK;
+                    final int sessionExpiryInterval = (int) mqttProperty.sessionExpiry();
```

Comment:
Perhaps let's call it `sessionExpiry` as `sessionExpiryInterval` is a bit of a mouthful and we already made this decision on the method name `.sessionExpiry()`.